File: 55-jump-game/55-jump-game.cpp
class Solution {
public:
    bool canJump(vector<int>& nums) {
        
        if (nums.size() == 1) return true;
        
        int len = nums.size();
        vector<int> dp(len, 0);
        dp[0] = nums[0];
        
        for (int i = 1; i < len; i++) {
            if (dp[i - 1] < i) return false;
            
            dp[i] = max(dp[i - 1], nums[i] + i);
            
            if (dp[i] >= len - 1) return true;
        }
        
        return true;
    }
};

// [2, 4, 0, 0, 0, 1 ]

File: 55-jump-game/55-jump-game.py
class Solution:
    def canJump(self, nums: List[int]) -> bool:
        
        dest = len(nums)
        pos = 0
        
        for i in range(0, dest):
            
            pos = max(i + nums[i], pos)
            
            if pos >= dest - 1: return True
            if pos <= i: return False
                
        return True